### ISSUE TITLE
Retrieve more data out of the training loop

### ISSUE THREAD

**Issue by Magellol** [Libre-YOLO/libreyolo#166]
Hey! I'm looking into hooking up training a YOLOX model to [mlflow](https://mlflow.org/) and so far I was only able to log the final metrics defined here https://github.com/LibreYOLO/libreyolo/blob/449f9781ed1e1c574f7bce6c4f4f8efdb0b54c54/libreyolo/training/trainer.py#L402-L411

`epoch_losses` already contains the loss curve. This is good, but I was curious if we could consider more data points like the LR curve, mAP50/mAP50-95 per epoch and so on. I don't know if this means returning these once the training session ends or exposing a hook we can tap into after each epoch. What do you think?

Happy to help implement anything!

**Comment by EHxuban11** [comment]
Hello @Magellol , thank you for opening this issue. 
To be honest, I have no idea about mlflow, and yeah, logging in general has received very little attention from our part, so it's good that you raise this point. 

I found a similar issue in the RF-DETR library with so many people requesting to add mlflow: https://github.com/roboflow/rf-detr/issues/110, which makes me think that it is an important feature to add to the library. Maybe we can read about how they do it in RF-DETR, just to understand how it's implemented in a mature repository.

They added support for mlflow in RF-DETR in the following PR: https://github.com/roboflow/rf-detr/pull/109
They also support others like wandb, mlflow and ClearML

I'm going to read how they do it and we can exchange ideas!

**Comment by Magellol** [comment]
Sounds good. I'll take on some inspirations as well.

On my end, since we can't tap onto the end of one epoch, I'm just logging everything at the end. 
It ends up looking something like this.

<img width="1508" height="992" alt="Image" src="https://github.com/user-attachments/assets/1c0a0fe9-9ea4-42e3-8fd2-c98d1c7521c8" />

In general, I think giving the low-level tools to tap into the training process would be more flexible and allow folks to leverage them for their favourite tracking platform without having to endlessly support many platforms. Curious to hear your toughts!

**Comment by EHxuban11** [comment]
Hello @Magellol , 

I was thinking to add a hook:
on_train_epoch_end -> emitted once per epoch, after train+val finished (if validation did not end, then validation is simply empty)

And this hookcan report:
epoch, total_epochs, train_loss, loss_components, lr, epoch_seconds, save_dir, model_family, val_metrics, is_best, best_metric, best_epoch

wdyt?

**Comment by Magellol** [comment]
Yeah I think that would be helpful. Does `loss_components` only contain validation loss components?

**Comment by EHxuban11** [comment]
@Magellol  I was thinking training loss componenets such ass the loss of that epoch and the individual losses that compose it.

**Comment by EHxuban11** [comment]
@Magellol 
Actually, only the loss components of that training epoch  which are going to be different per model but could be for instance iou loss or class lost

**Comment by EHxuban11** [comment]
Hello, I'm gonna start woring on this today because I really need it for a project.

**Comment by Magellol** [comment]
Cool! Feel free to tag me in the PR, it'll be good for me to get more exposure.

**Comment by EHxuban11** [comment]
Hello @Magellol, sry for being a bit ausent. I'm preparing the library for the v1.2.0 release which +100 people will use in the cursor Madrid hackathon where the main track is about using LibreYOLO 💀💀💀 and I need to make sure that everything works on windows / linux / mac for the flagship models: YOLO9 and RF-DETR

Okey so a first proposal of your request has been merged to dev for the flagship models, YOLO9 and RF-DETR.  Feel free to test it. It has hooks like you asked. I can give you some documentation generate on the fligh by claude regarding how to use this hooks. If you like how they work maybe you can extend them to YOLOX or other models. If there is something you want to improve or change feel free to comment here or open a PR.

Example code from claude code: 
```
 from libreyolo import LibreYOLO9
  from libreyolo.training import TrainEpochEvent

  # Anything callable taking a TrainEpochEvent works.
  # It fires once per epoch, after train + val.

  def on_epoch(e: TrainEpochEvent):
      print(
          f"epoch {e.epoch+1}/{e.total_epochs}  "
          f"loss={e.train_loss:.4f}  "
          f"lr={list(e.lr.values())[0]:.2e}  "
          f"mAP50={e.val_metrics.get('mAP50', float('nan')):.3f}  "
          f"mAP50-95={e.val_metrics.get('mAP50_95', float('nan')):.3f}  "
          f"{'(best)' if e.is_best else ''}"
      )

  model = LibreYOLO9("yolo9-s.pt")
  model.train(data="coco8.yaml", epochs=10, callbacks=on_epoch)

  And the TrainEpochEvent fields he can pull from, so he can build whatever he wants (matplotlib curve, CSV, MLflow, W&B, Slack ping, …):

  e.epoch, e.total_epochs                # ints
  e.train_loss                            # float
  e.train_loss_items                      # {"iou": ..., "cls": ..., ...} (per-component)
  e.lr                                    # {"group0": ..., "group1": ...}  (per param-group)
  e.val_metrics                           # {"mAP50": ..., "mAP50_95": ..., ...}
  e.is_best, e.best_metric, e.best_epoch  # best-so-far tracking
  e.epoch_seconds                         # wall time of this epoch
  e.model_family, e.model_size, e.task, e.save_dir
```

**Comment by Magellol** [comment]
Hey so i've taken the callback for a spin. I think it makes sense, but I will play with it more and plug it to an actual use case.
Quick question though, you mention enabling it for YOLO9 and RF-DETR but I've tried with YOLOX and it also considers the callback and the metrics are logged. Did I miss anything?

EDIT: ah, it looks like since `YOLOXTrainer` inherits from `BaseTrainer` and the fact the `kwargs` are piped through `BaseTrainer`, it works out of the box.

**Comment by EHxuban11** [comment]
Hey, @Magellol , yeah it looks like happy accident XD. But we can definetely make sure that it works 100% for every model in the future. Let me know if you have any improvement suggestion so that the base is good before we lock in to this solution.

In addition to fully covering all the models, in the near future we can work on a layer above the hooks to integrate with Tensorboard, mlflow, etc. Both the hooks and integrations are very important for the future of the library :)

**Comment by Magellol** [comment]
My use case is rather simple but I can't think of anything that I would be missing so yet. I've added a few screenshot from my mlflow instance below.

If you agree, I think we can lock in on this implementation and we can always iterate on later?

To enable it officially for other models I think it would just be a matter of adding the missing `callbacks=None` property to the `train` argument of the various `Libre` model wrapper 🤔 
https://github.com/LibreYOLO/libreyolo/blob/74013a6d6336122df35c7c02047bc7cb045ab4e8/libreyolo/models/yolo9/model.py#L270

This is how I ended up plugging to `mlflow` for any future reference

```python
def on_epoch(e: TrainEpochEvent):
    mlflow.log_metric("train/loss", e.train_loss, step=e.epoch)
    for name, value in e.train_loss_items.items():
        mlflow.log_metric(f"train/loss/{name}", value, step=e.epoch)

    mlflow.log_metric("lr", next(iter(e.lr.values())), step=e.epoch)
    mlflow.log_metric("epoch_seconds", e.epoch_seconds, step=e.epoch)
    if e.validated:
        mlflow.log_metric(
            "val_mAP50", e.val_metrics["metrics/mAP50"], step=e.epoch
        )
        mlflow.log_metric(
            "val_mAP50_95",
            e.val_metrics["metrics/mAP50-95"],
            step=e.epoch,
        )
```

<img width="3134" height="1309" alt="Image" src="https://github.com/user-attachments/assets/97c826cc-48cd-4254-bf6f-7b3b1fd838ed" />

<img width="3193" height="1124" alt="Image" src="https://github.com/user-attachments/assets/30a3134a-c7ef-4af5-8301-c388391c43d8" />

**Comment by EHxuban11** [comment]
Hey @Magellol , 

Sorry for taking a lot of time to answer, again... I've been super busy with the LibreYOLO hackathon that happened earlier this week.

I'm happy that you like the base, then we can move forward with this solution and expand it to the other models.

Would you like to contribute in adding proper support for the hooks to the other models? If not, don't worry, I can work on it when I have time because this is kind of important.

**Comment by Magellol** [comment]
Sounds good. Yeah, I can try to take a look at it this week.
